torch/sparse/triton_ops/triton_bsr_dense_mm.py

```diff
@@ -0,0 +1,295 @@
+    # That will be advanced in the loop according to values in col_indices.
+    dense_block_ptrs = (
+        dense_ptr
+        + dense_batch_stride * batch_idx
+        + dense_tiled_col_stride * col_block_pid
+        + dense_row_block_stride * col_block_arange[:, None]
+        + dense_col_block_stride * row_block_arange[None, :]
+    )
+    # Pointers are set to exact write-to locations
+    output_ptrs = (
+        output_ptr
+        + output_batch_stride * batch_idx
+        + output_tiled_row_stride * row_idx
+        + output_tiled_col_stride * col_block_pid
+        + output_row_block_stride * row_block_arange[:, None]
+        + output_col_block_stride * row_block_arange[None, :]
+    )
+
+    output_acc_block = tl.zeros((BLOCKSIZE_ROW, BLOCKSIZE_ROW), tl.float32)
+    col_index_nnz_ptr = col_indices_ptr + row_idx_nnz_offset * col_indices_stride
+    for nnz_idx in range(row_idx_nnz):
+        values_block = tl.load(values_block_ptrs)
+
+        # find which row of dense needs to get loaded
+        # for multiplication with values_block.
+        dense_row_idx = tl.load(col_index_nnz_ptr)
+        dense_block = tl.load(dense_block_ptrs + dense_tiled_row_stride * dense_row_idx)
+
+        # do block mm
+        output_acc_block += tl.dot(values_block, dense_block)
+
+        # move val ptrs to the next block in the row
+        values_block_ptrs += values_nnz_stride
+        col_index_nnz_ptr += col_indices_stride
+
+    # write back the result
+    tl.store(output_ptrs, output_acc_block.to(output_ptr.dtype.element_ty))
+
+
+def bsr_dense_mm(bsr, dense):
+    # TODO: insert checks
+    m, k = bsr.shape[-2:]
+    k1, n = dense.shape[-2:]
+
+    # TODO: probably make sure that inputs are properly contiguous
+
+    # Required to undo the fake batch dimension insertion.
+    original_batch_dims_broadcasted = torch.broadcast_shapes(
+        bsr.shape[:-2], dense.shape[:-2]
+    )
+
+    # Short circuit if lhs is zero
+    if bsr._nnz() == 0:
+        return torch.zeros(
+            original_batch_dims_broadcasted + (m, n),
+            dtype=dense.dtype,
+            device=dense.device,
+        )
+
+    # Introduce fake batch dimension if not present for convenience.
+    def unsqueeze_batch_dim(t, n_non_batch_dims):
+        if t.dim() > n_non_batch_dims:
+            return t
+        else:
+            return t.unsqueeze(0)
+
+    crow_indices = unsqueeze_batch_dim(bsr.crow_indices(), 1)
+    col_indices = unsqueeze_batch_dim(bsr.col_indices(), 1)
+    values = unsqueeze_batch_dim(bsr.values(), 3)
+    dense = unsqueeze_batch_dim(dense, 2)
+    nnz = values.shape[-3]
+    blocksize = values.shape[-2:]
+
+    # Compute broadcasted batch dimension
+    bsr_batch_dims = values.shape[:-3]
+    dense_batch_dims = dense.shape[:-2]
+    batch_dims_broadcasted = torch.broadcast_shapes(bsr_batch_dims, dense_batch_dims)
+
+    # Allocate output
+    output = torch.zeros(
```

Comment:
nit: can we use empty?

Reply:
No, we cannot, unless we know that each block of output is gonna get hit, otherwise they should be zeroed out.